aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu

```diff
@@ -33,6 +35,82 @@ using namespace at::sparse_csr;
 // certain utiliy functions are usable from sparse COO.
 using namespace at::sparse;
 
+Tensor& addmm_out_sparse_csr_dense_cuda(
+  const Tensor& self,
+  const SparseCsrTensor& sparse,
+  const Tensor& dense,
+  const Scalar& beta,
+  const Scalar& alpha,
+  Tensor& r)
+{
+
+  TORCH_INTERNAL_ASSERT(sparse.is_sparse_csr());
+  Tensor t = *expand_size(self, {sparse.size(0), dense.size(1)}, "addmm_out_sparse_csr");
+
+  TORCH_INTERNAL_ASSERT(t.device().type() == kCUDA);
+  TORCH_CHECK(
+      r.device().type() == kCUDA,
+      "addmm: expected 'out' to be CUDA tensor, but got CUDA tensor");
+  TORCH_CHECK(
+      sparse.device().type() == kCUDA,
+      "addmm: expected 'mat1' to be a CUDA tensor, but got a CUDA tensor");
+  TORCH_CHECK(
+      dense.device().type() == kCUDA,
+      "addmm: expected 'mat2' to be a CUDA tensor, but got a CUDA tensor");
+
+  TORCH_CHECK(
+      sparse.dim() == 2,
+      "addmm: 2-D matrices expected, got ",
+      sparse.dim(),
+      "D tensor");
+  TORCH_CHECK(
+      dense.dim() == 2,
+      "addmm: 2-D matrices expected, got ",
+      dense.dim(),
+      "D tensor");
+
+  TORCH_CHECK(
+      r.is_contiguous(),
+      "out argument must be contiguous, but got: ",
+      r.suggest_memory_format());
+
+  // mxk * kxn = mxn
+  int64_t m = sparse.size(0);
+  int64_t k = sparse.size(1);
+  int64_t n = dense.size(1);
+
+  TORCH_CHECK(
+      dense.size(0) == k,
+      "addmm: Expected dense matrix (dense) size(0)=",
+      k,
+      ", got ",
+      dense.size(0));
+  TORCH_CHECK(
+      sparse.size(1) == k,
+      "addmm: Expected sparse matrix (sparse) size(1)=",
+      k,
+      ", got ",
+      sparse.size(1));
+  resize_output(r, {m, n});
+
+  if (sparse.crow_indices().scalar_type() != kInt) {
+    TORCH_WARN(
+        "Pytorch is compiled with MKL LP64 and will convert crow_indices to int32.");
+  }
+  if (sparse.col_indices().scalar_type() != kInt) {
+    TORCH_WARN(
+        "Pytorch is compiled with MKL LP64 and will convert col_indices to int32.");
```

Comment:
Thanks Pearu for the comment. Unfortunately, cuSparse/csrmm2 doesn't support int64 indices. 

https://github.com/pytorch/pytorch/blob/e4be80c1b82a907fd1c799624af5987ba1e843b3/aten/src/ATen/native/sparse/cuda/SparseCUDABlas.cu#L200

I will remove these warnings to be consistent with .
https://github.com/pytorch/pytorch/blob/e4be80c1b82a907fd1c799624af5987ba1e843b3/aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu#L62

Reply:
> Thanks Pearu for the comment. Unfortunately, cuSparse/csrmm2 doesn't support int64 indices.

See https://github.com/pytorch/pytorch/blob/e4be80c1b82a907fd1c799624af5987ba1e843b3/aten/src/ATen/native/sparse/cuda/SparseCUDABlas.cu#L125-L126

IIUC, the task boils down to generalizing `template<typename T> void _csrmm2...` to `template<typename T, typename I>
void _csrmm2` where `I` is index type.

If this makes sense, add `TODO` comment for follow-up work.

Reply:
Sure, make sense.

Reply:
I will create an issue for this too as it could affect other things.  
NVIDIA docs mention that CUSPARSE_INDEX_64I can be used but for CUSPARSE_SPMM_CSR_ALG2/CUSPARSE_SPMM_CSR_ALG3 currently is it uses CUSPARSE_CSRMM_ALG1
https://docs.nvidia.com/cuda/cusparse/index.html

> cusparseSpMM supports the following index type for representing the sparse matrix matA:
> 32-bit indices (CUSPARSE_INDEX_32I)
> 64-bit indices (CUSPARSE_INDEX_64I) only with CUSPARSE_SPMM_COO_ALG4, CUSPARSE_SPMM_CSR_ALG2, and CUSPARSE_SPMM_CSR_ALG3 algorithms

Reply:
Mentioned here: https://github.com/pytorch/pytorch/issues/58899